File: notify.py
"""
notify.py
---------
Polls Google Calendar for recently ended meetings, matches attendees to
Attio deals, and sends Slack notifications with action buttons.

Usage:
    python notify.py              # Normal run
    python notify.py --dry-run    # Show what would be sent without sending
    python notify.py --window 15  # Check last 15 minutes instead of default 10
"""

import argparse
import json
import os
from datetime import datetime, timezone
from pathlib import Path

from attio_helpers import (
    extract_domain,
    find_person_by_email,
    find_company_by_domain,
    find_deals_by_person,
    find_deals_by_company,
    get_deal_details,
    get_person_details,
    get_company_name,
    IGNORED_DOMAINS,
)
from gcal_helpers import get_recently_ended_meetings
from slack_helpers import send_founder_call_message
from errors import logger, ErrorCollector

STATE_FILE = Path(__file__).parent / "notified_events.json"
PRUNE_DAYS = 30


# ── State management ─────────────────────────────────────────────────────────

def load_state() -> dict:
    """Load the notified events state file."""
    if STATE_FILE.exists():
        try:
            return json.loads(STATE_FILE.read_text())
        except (json.JSONDecodeError, OSError) as exc:
            logger.warning("Failed to load state file: %s", exc)
    return {}


def save_state(state: dict):
    """Save the notified events state file, pruning old entries."""
    cutoff = datetime.now(timezone.utc).timestamp() - (PRUNE_DAYS * 86400)
    pruned = {
        k: v for k, v in state.items()
        if datetime.fromisoformat(v).timestamp() > cutoff
    }
    STATE_FILE.write_text(json.dumps(pruned, indent=2))
    logger.info("State saved (%d entries, pruned %d old)", len(pruned), len(state) - len(pruned))


def state_key(event_id: str, deal_record_id: str) -> str:
    return f"{event_id}:{deal_record_id}"


# ── Deal matching ─────────────────────────────────────────────────────────────

def match_attendees_to_deals(attendee_emails: list[str]) -> list[dict]:
    """Match meeting attendees to Attio deals.

    Returns a list of dicts, each with:
      - deal: deal details dict
      - contact_email: the matched attendee email
      - contact_name: the matched person's name (if found)
      - company_name: the matched company name
      - match_type: "person" or "company"
    """
    matches = []
    seen_deal_ids = set()

    for email in attendee_emails:
        # Try matching via person → deal (primary_contact)
        person_ids = find_person_by_email(email)
        for person_id in person_ids:
            deals = find_deals_by_person(person_id)
            for deal in deals:
                if deal["record_id"] in seen_deal_ids:
                    continue
                seen_deal_ids.add(deal["record_id"])

                person_info = get_person_details(person_id)
                deal_details = get_deal_details(deal["record_id"])
                company_name = ""
                if deal_details and deal_details.get("company_record_id"):
                    company_name = get_company_name(deal_details["company_record_id"])

                matches.append({
                    "deal": deal_details or {"record_id": deal["record_id"]},
                    "contact_email": email,
                    "contact_name": person_info["name"] if person_info else "",
                    "company_name": company_name,
                    "match_type": "person",
                })

        # Try matching via domain → company → deal
        domain = extract_domain(email)
        if not domain or domain in IGNORED_DOMAINS:
            continue

        company_ids = find_company_by_domain(domain)
        for company_id in company_ids:
            deals = find_deals_by_company(company_id)
            for deal in deals:
                if deal["record_id"] in seen_deal_ids:
                    continue
                seen_deal_ids.add(deal["record_id"])

                deal_details = get_deal_details(deal["record_id"])
                company_name = get_company_name(company_id)

                matches.append({
                    "deal": deal_details or {"record_id": deal["record_id"]},
                    "contact_email": email,
                    "contact_name": "",
                    "company_name": company_name,
                    "match_type": "company",
                })

    return matches


# ── Main ──────────────────────────────────────────────────────────────────────

def run(dry_run: bool = False, window_minutes: int = 10):
    """Main entry point: poll calendar, match deals, send Slack messages."""
    errors = ErrorCollector()
    state = load_state()

    logger.info("Starting notification run (dry_run=%s, window=%dm)", dry_run, window_minutes)

    # 1. Get recently ended meetings
    try:
        meetings = get_recently_ended_meetings(minutes_ago=window_minutes)
    except Exception as exc:
        errors.add("calendar_poll", exc)
        logger.error("Failed to poll calendar: %s", exc)
        return

    if not meetings:
        logger.info("No recently ended meetings found.")
        return

    logger.info("Found %d recently ended meetings", len(meetings))
    notifications_sent = 0

    # 2. For each meeting, match attendees to deals
    for meeting in meetings:
        try:
            logger.info("Processing meeting: %s (%s)", meeting["title"], meeting["event_id"])
            matches = match_attendees_to_deals(meeting["attendee_emails"])

            if not matches:
                logger.info("  No deal matches for meeting '%s'", meeting["title"])
                continue

            logger.info("  Found %d deal match(es)", len(matches))

            # 3. Send Slack notification for each matched deal
            for match in matches:
                deal = match["deal"]
                key = state_key(meeting["event_id"], deal.get("record_id", ""))

                if key in state:
                    logger.info("  Already notified for %s, skipping", key)
                    continue

                deal_info = {
                    "record_id": deal.get("record_id", ""),
                    "deal_name": deal.get("deal_name", "Unknown"),
                    "company_name": match["company_name"],
                    "contact_name": match["contact_name"],
                    "contact_email": match["contact_email"],
                }

                if dry_run:
                    logger.info(
                        "  [DRY RUN] Would send Slack message: deal=%s company=%s contact=%s",
                        deal_info["deal_name"],
                        deal_info["company_name"],
                        deal_info["contact_email"],
                    )
                else:
                    try:
                        send_founder_call_message(deal_info, meeting)
                        notifications_sent += 1
                    except Exception as exc:
                        errors.add(f"slack_send:{deal.get('record_id', '')}", exc)
                        continue

                state[key] = datetime.now(timezone.utc).isoformat()

        except Exception as exc:
            errors.add(f"meeting:{meeting.get('event_id', 'unknown')}", exc)
            continue

    # 4. Save state and report
    if not dry_run:
        save_state(state)

    logger.info("Run complete. Notifications sent: %d", notifications_sent)
    if errors.has_errors:
        logger.warning("Run completed with errors:\n%s", errors.summary())


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Post-call deal notification system.")
    parser.add_argument("--dry-run", action="store_true",
                        help="Log what would be sent without sending")
    parser.add_argument("--window", type=int, default=10,
                        help="Minutes to look back for ended meetings (default: 10)")
    args = parser.parse_args()

    run(dry_run=args.dry_run, window_minutes=args.window)

File: src/notify.ts
import "dotenv/config";
import { extractDomain, IGNORED_DOMAINS } from "./types/crm.js";
import type { CRM } from "./types/crm.js";
import type { Recording } from "./types/recording.js";
import { createAttioCRM } from "./lib/attio.js";
import { createFathomRecording } from "./lib/fathom.js";
import { getRecentlyEndedMeetings, type CalendarMeeting } from "./lib/gcal.js";
import { getFathomSummary } from "./lib/gmail.js";
import { sendZoeMessage } from "./lib/slack.js";
import { summarizeTranscript } from "./lib/ai.js";
import { loadNotifiedEvents, saveNotifiedEvents, stateKey, readLastNotify, writeLastNotify } from "./lib/state.js";
import { ErrorCollector, createGitHubIssue, logger } from "./lib/errors.js";

interface DealMatch {
  dealRecordId: string;
  dealName: string;
  companyName: string;
  contactEmail: string;
  contactName: string;
  companyRecordId: string;
}

interface MatchDiagnostic {
  email: string;
  personFound: boolean;
  personIds: string[];
  dealsByPerson: number;
  dealsFilteredByStage: number;
  domain: string | null;
  domainIgnored: boolean;
  companyFound: boolean;
  companyIds: string[];
  dealsByCompany: number;
}

interface MatchResult {
  matches: DealMatch[];
  diagnostics: MatchDiagnostic[];
}

const EXCLUDED_STAGES = new Set(["Reject", "Invested"]);

async function matchAttendeesToDeals(
  crm: CRM,
  attendeeEmails: string[],
): Promise<MatchResult> {
  const matches: DealMatch[] = [];
  const diagnostics: MatchDiagnostic[] = [];
  const seenDealIds = new Set<string>();

  for (const email of attendeeEmails) {
    const diag: MatchDiagnostic = {
      email,
      personFound: false,
      personIds: [],
      dealsByPerson: 0,
      dealsFilteredByStage: 0,
      domain: extractDomain(email),
      domainIgnored: false,
      companyFound: false,
      companyIds: [],
      dealsByCompany: 0,
    };

    // Person → deal
    const personIds = await crm.findPersonByEmail(email);
    diag.personFound = personIds.length > 0;
    diag.personIds = personIds;

    for (const personId of personIds) {
      const deals = await crm.findDealsByPerson(personId);
      diag.dealsByPerson += deals.length;

      for (const deal of deals) {
        if (seenDealIds.has(deal.recordId)) continue;
        seenDealIds.add(deal.recordId);

        const person = await crm.getPersonDetails(personId);
        const details = await crm.getDealDetails(deal.recordId);

        if (details && EXCLUDED_STAGES.has(details.dealStage)) {
          logger.info(`  Skipping deal ${details.dealName} — stage "${details.dealStage}" excluded`);
          diag.dealsFilteredByStage++;
          continue;
        }

        let companyName = "";
        if (details?.companyRecordId) {
          companyName = await crm.getCompanyName(details.companyRecordId);
        }

        matches.push({
          dealRecordId: deal.recordId,
          dealName: details?.dealName ?? deal.dealName ?? "Unknown",
          companyName,
          contactEmail: email,
          contactName: person?.name ?? "",
          companyRecordId: details?.companyRecordId ?? "",
        });
      }
    }

    // Domain → company → deal
    const domain = diag.domain;
    if (!domain || IGNORED_DOMAINS.has(domain)) {
      diag.domainIgnored = !domain || IGNORED_DOMAINS.has(domain);
      diagnostics.push(diag);
      continue;
    }

    const companyIds = await crm.findCompanyByDomain(domain);
    diag.companyFound = companyIds.length > 0;
    diag.companyIds = companyIds;

    for (const companyId of companyIds) {
      const deals = await crm.findDealsByCompany(companyId);
      diag.dealsByCompany += deals.length;

      for (const deal of deals) {
        if (seenDealIds.has(deal.recordId)) continue;
        seenDealIds.add(deal.recordId);

        const details = await crm.getDealDetails(deal.recordId);

        if (details && EXCLUDED_STAGES.has(details.dealStage)) {
          logger.info(`  Skipping deal ${details.dealName} — stage "${details.dealStage}" excluded`);
          diag.dealsFilteredByStage++;
          continue;
        }

        const companyName = await crm.getCompanyName(companyId);

        matches.push({
          dealRecordId: deal.recordId,
          dealName: details?.dealName ?? deal.dealName ?? "Unknown",
          companyName,
          contactEmail: email,
          contactName: "",
          companyRecordId: companyId,
        });
      }
    }

    diagnostics.push(diag);
  }

  return { matches, diagnostics };
}

async function getCallSummary(
  recording: Recording,
  meetingTitle: string,
  attendeeEmails: string[],
  companyName: string,
): Promise<{ summary: string; fathomLink: string }> {
  // Fetch only today's Fathom meetings
  const todayStart = new Date();
  todayStart.setHours(0, 0, 0, 0);
  const meetings = await recording.fetchMeetings(todayStart.toISOString());

  // Match strategy: attendee email first, then title fallback
  let fathomMeeting = meetings.find((m) =>
    m.invitees.some((inv) =>
      attendeeEmails.some((email) => inv.email.toLowerCase() === email.toLowerCase()),
    ),
  );

  // Fallback to title match
  if (!fathomMeeting) {
    const titleLower = meetingTitle.toLowerCase();
    fathomMeeting = meetings.find((m) => {
      const ft = m.title.toLowerCase();
      return ft.includes(titleLower) || titleLower.includes(ft);
    });
  }

  logger.info(
    `Fathom match for "${meetingTitle}": ${fathomMeeting ? `found "${fathomMeeting.title}" (id: ${fathomMeeting.id})` : "not found"} (searched ${meetings.length} meetings today)`,
  );

  let fathomLink = "";
  let summaryText = "";

  if (fathomMeeting) {
    fathomLink = fathomMeeting.shareUrl;

    // Use Fathom's default_summary first (already processed by Fathom)
    if (fathomMeeting.defaultSummary) {
      summaryText = fathomMeeting.defaultSummary;
      logger.info(`Using Fathom default_summary for "${meetingTitle}"`);
    }
    // Then try inline transcript → Claude
    else if (fathomMeeting.transcript) {
      summaryText = await summarizeTranscript(fathomMeeting.transcript, companyName);
    }
    // Fetch transcript from dedicated endpoint if not in list response
    else {
      const transcript = await recording.getTranscript(fathomMeeting.id);
      if (transcript) {
        logger.info(`Fetched transcript for "${meetingTitle}" via getTranscript`);
        summaryText = await summarizeTranscript(transcript, companyName);
      }
    }
  }

  // Fallback to Gmail Fathom email → Claude
  if (!summaryText) {
    const attendeeEmail = attendeeEmails[0] ?? "";
    const gmailText = await getFathomSummary(meetingTitle, attendeeEmail, companyName);
    if (gmailText) {
      summaryText = await summarizeTranscript(gmailText, companyName);
    }
  }

  return { summary: summaryText, fathomLink };
}

async function buildLinkedinMap(
  crm: CRM,
  emails: string[],
): Promise<Record<string, { name: string; linkedin: string }>> {
  const map: Record<string, { name: string; linkedin: string }> = {};
  for (const email of emails) {
    const personIds = await crm.findPersonByEmail(email);
    if (!personIds.length) continue;
    const details = await crm.getPersonDetails(personIds[0]);
    const linkedin = await crm.getPersonLinkedin(personIds[0]);
    if (details) {
      map[email] = { name: details.name, linkedin };
    }
  }
  return map;
}

async function run(dryRun: boolean, fallbackWindowMinutes: number) {
  const errors = new ErrorCollector();
  const state = loadNotifiedEvents();
  const crm = createAttioCRM();
  const recording = createFathomRecording();

  // Use last notify timestamp for window, fall back to fixed window
  const lastNotify = readLastNotify();
  const now = new Date();
  let windowMinutes: number;

  if (lastNotify) {
    const elapsed = Math.ceil((now.getTime() - new Date(lastNotify).getTime()) / 60_000);
    // Cap at 180 min to avoid scanning too far back (e.g. overnight)
    windowMinutes = Math.min(Math.max(elapsed, fallbackWindowMinutes), 180);
  } else {
    windowMinutes = fallbackWindowMinutes;
  }

  logger.info(`Starting notification run (dry_run=${dryRun}, window=${windowMinutes}m, last_notify=${lastNotify ?? "none"})`);

  // Save the check timestamp now so the next run covers from this point
  if (!dryRun) writeLastNotify(now.toISOString());

  let meetings: CalendarMeeting[];
  try {
    meetings = await getRecentlyEndedMeetings(windowMinutes);
  } catch (err) {
    errors.add("calendar_poll", err);
    return;
  }

  if (!meetings.length) {
    logger.info("No recently ended meetings found.");
    return;
  }

  logger.info(`Found ${meetings.length} recently ended meetings`);
  let notificationsSent = 0;

  for (const meeting of meetings) {
    try {
      logger.info(`Processing meeting: ${meeting.title} (${meeting.eventId})`);
      const { matches, diagnostics } = await matchAttendeesToDeals(crm, meeting.attendeeEmails);

      if (!matches.length) {
        const allFilteredByStage = diagnostics.some((d) => d.dealsFilteredByStage > 0)
          && diagnostics.every((d) => d.dealsByPerson === d.dealsFilteredByStage && d.dealsByCompany === 0 || d.dealsByCompany === d.dealsFilteredByStage);

        if (allFilteredByStage) {
          logger.info(`  All deals filtered by stage for meeting '${meeting.title}' — skipping`);
          continue;
        }

        logger.info(`  No deal matches for meeting '${meeting.title}'`);
        const unmatchedKey = stateKey(meeting.eventId, "unmatched");
        if (unmatchedKey in state) {
          logger.info(`  Already filed unmatched issue for ${meeting.eventId}, skipping`);
          continue;
        }
        if (!dryRun) {
          const diagLines = diagnostics.map((d) => {
            const steps: string[] = [`  - \`${d.email}\``];
            if (d.personFound) {
              steps.push(`    - Person found (${d.personIds.join(", ")})`);
              steps.push(`    - Deals via primary_contact: ${d.dealsByPerson}`);
            } else {
              steps.push(`    - No person found in Attio`);
            }
            if (d.domainIgnored) {
              steps.push(`    - Domain \`${d.domain ?? "none"}\` is personal/ignored — company lookup skipped`);
            } else if (d.domain) {
              if (d.companyFound) {
                steps.push(`    - Company found for \`${d.domain}\` (${d.companyIds.join(", ")})`);
                steps.push(`    - Deals via company: ${d.dealsByCompany}`);
              } else {
                steps.push(`    - No company found for \`${d.domain}\``);
              }
            }
            return steps.join("\n");
          });

          await createGitHubIssue(
            `Unmatched call: ${meeting.title}`,
            [
              `**Meeting:** ${meeting.title}`,
              `**Ended:** ${meeting.endTime}`,
              `**Attendees:** ${meeting.attendeeEmails.join(", ")}`,
              "",
              "### Match pipeline results",
              "",
              ...diagLines,
            ].join("\n"),
            ["unmatched-call"],
          );
          state[unmatchedKey] = new Date().toISOString();
        }
        continue;
      }

      logger.info(`  Found ${matches.length} deal match(es)`);

      for (const match of matches) {
        const key = stateKey(meeting.eventId, match.dealRecordId);
        if (key in state) {
          logger.info(`  Already notified for ${key}, skipping`);
          continue;
        }

        if (dryRun) {
          logger.info(`  [DRY RUN] Would send: deal=${match.dealName} company=${match.companyName}`);
        } else {
          try {
            // Get transcript + summary + fathom link
            const { summary, fathomLink } = await getCallSummary(
              recording,
              meeting.title,
              meeting.attendeeEmails,
              match.companyName,
            );

            // Build LinkedIn map for attendees
            const linkedinMap = await buildLinkedinMap(crm, meeting.attendeeEmails);

            await sendZoeMessage({
              dealRecordId: match.dealRecordId,
              dealName: match.dealName,
              companyName: match.companyName,
              contactEmail: match.contactEmail,
              contactName: match.contactName,
              summary,
              fathomLink,
              attendeeEmails: meeting.attendeeEmails,
              linkedinMap,
              eventId: meeting.eventId,
              meetingTitle: meeting.title,
              endTime: meeting.endTime,
            });

            notificationsSent++;
          } catch (err) {
            errors.add(`slack_send:${match.dealRecordId}`, err);
            continue;
          }
        }

        state[key] = new Date().toISOString();
      }
    } catch (err) {
      errors.add(`meeting:${meeting.eventId}`, err);
    }
  }

  if (!dryRun) saveNotifiedEvents(state);

  logger.info(`Run complete. Notifications sent: ${notificationsSent}`);
  if (errors.hasErrors) {
    logger.warn(`Run completed with errors:\n${errors.summary()}`);
  }
}

async function main() {
  const args = process.argv.slice(2);
  const dryRun = args.includes("--dry-run");
  const windowIdx = args.indexOf("--window");
  const window = windowIdx >= 0 ? parseInt(args[windowIdx + 1], 10) : 10;

  await run(dryRun, window);
}

main().catch((err) => {
  logger.error("Fatal error:", err);
  process.exit(1);
});
